### PR TITLE
refactor(core): make di error messages tree shakable

### PR DESCRIPTION
**Pull request by ramthir**
Tree shake error messages from the production build and introduce error codes.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?
Errors under DI are not tree-shakable in the production build.

Issue Number: N/A

## What is the new behavior?
Make the Errors under DI tree-shakable in the production build.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/di/reflective_key.ts [hunk 1]
```diff
@@ -6,7 +6,9 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {RuntimeError, RuntimeErrorCode} from '../errors';
 import {stringify} from '../util/stringify';
+
```
Comment (pkozlowski-opensource) [inline packages/core/src/di/reflective_key.ts:11]: nit: revert formatting?

File: packages/core/src/di/reflective_key.ts [hunk 2]
```diff
@@ -35,7 +37,7 @@ export class ReflectiveKey {
    */
   constructor(public token: Object, public id: number) {
     if (!token) {
-      throw new Error('Token must be defined!');
+      throw new RuntimeError(RuntimeErrorCode.MISSING_INJECTION_TOKEN, 'Token must be defined!');
```
Comment (pkozlowski-opensource) [inline packages/core/src/di/reflective_key.ts:40]: Shouldn't we guard the "Token must be defined!" error message with `ngDevMode`?

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
@ramthir thnx for the PR. It seems like the error message in the `ReflectiveKey` class is not guarded with `ngDevMode` - is it intentional? 

Also the commit message should be changed from `refactor(common): ...` to `refactor(core): ...` as the changes touch the `core` package and not the `common` one.

File: packages/core/src/di/reflective_key.ts (hunk 1 above)
Comment (ramthir) [reply packages/core/src/di/reflective_key.ts:11]: Storage, it gets formatted with `yarn ng-dev format changed`. I've reverted it.

**Comment by ramthir** [comment]
@pkozlowski-opensource, thanks a lot for the review comments. I've made the changes you requested.

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM

Reviewed-for: public-api

**Review by pkozlowski-opensource — APPROVED** [review]

**Review by atscott — APPROVED** [review]
reviewed-for: public-api

File: packages/core/src/di/reflective_key.ts [hunk 3]
```diff
@@ -35,7 +37,8 @@ export class ReflectiveKey {
    */
   constructor(public token: Object, public id: number) {
     if (!token) {
-      throw new Error('Token must be defined!');
+      throw new RuntimeError(
+          RuntimeErrorCode.INJECTION_TOKEN_UNDEFINED, ngDevMode && 'Token must be defined!');
```
Comment (AndrewKushnir) [inline packages/core/src/di/reflective_key.ts:41]: I think we can name this error code `MISSING_INJECTION_TOKEN`, similar to an existing `MISSING_INJECTION_CONTEXT` token:
```suggestion
          RuntimeErrorCode.MISSING_INJECTION_TOKEN, ngDevMode && 'Token must be defined!');
```

Comment (ramthir) [reply packages/core/src/di/reflective_key.ts]: @AndrewKushnir, I've changed the error code to `MISSING_INJECTION_TOKEN` so it's more aligned with the existing ones.

**Review by AndrewKushnir — APPROVED** [review]
@ramthir thanks for the update 👍

**Review by AndrewKushnir — APPROVED** [review]
Reviewed-for: public-api

**Comment by dylhunn** [comment]
Would you be able to rebase this @ramthir?

**Review by AndrewKushnir — APPROVED** [review]

**Comment by ramthir** [comment]
@dylhunn, I've rebased it. Now you can proceed with the merge.

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit 07828528be7cd64f0a8a9be3d574ab560815520d.

**Pull request closed without merging**